lib/auth/stableunixusers/stableunixusers.go

```diff
@@ -0,0 +1,292 @@
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindStableUNIXUser, types.VerbCreate, types.VerbRead); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	uid, err := s.obtainUIDForUsernameCached(ctx, req.GetUsername())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &stableunixusersv1.ObtainUIDForUsernameResponse{
+		Uid: uid,
+	}, nil
+}
+
+// obtainUIDForUsernameCached calls [*server.obtainUIDForUsernameUncached]
+// through the UID FnCache.
+func (s *server) obtainUIDForUsernameCached(ctx context.Context, username string) (int32, error) {
+	uid, err := utils.FnCacheGet(ctx, s.uidCache, username, func(ctx context.Context) (int32, error) {
+		ctx, cancel := context.WithTimeout(ctx, uidCacheTTL)
+		defer cancel()
+		return s.obtainUIDForUsernameUncached(ctx, username)
+	})
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return uid, nil
+}
+
+// obtainUIDForUsernameUncached reads or creates the stable UID for the given username.
+func (s *server) obtainUIDForUsernameUncached(ctx context.Context, username string) (int32, error) {
+	if username == "" {
+		return 0, trace.BadParameter("username must not be empty")
+	}
+
+	// we should only ever race with different auth servers on the same cluster,
+	// since we have a semaphore for the local auth
+	const maxAttempts = 3
+	for attempt := range maxAttempts {
+		if attempt > 0 {
+			select {
+			case <-ctx.Done():
+				return 0, trace.Wrap(context.Cause(ctx))
+			case <-time.After(time.Duration(attempt) * 100 * time.Millisecond):
+			}
+		}
+
+		uid, err := s.stableUNIXUsers.GetUIDForUsername(ctx, username)
+		if err == nil {
+			// TODO(espadolini): _potentially_ emit an audit log event with
+			// username and UID (it might spam the audit log unnecessarily)
+			return uid, nil
+		}
+		if !trace.IsNotFound(err) {
+			return 0, trace.Wrap(err)
+		}
+
+		var authPref readonly.AuthPreference
+		if attempt == 0 {
+			ap, err := s.readOnlyCache.GetReadOnlyAuthPreference(ctx)
+			if err != nil {
+				return 0, trace.Wrap(err)
+			}
+			authPref = ap
+		} else {
+			ap, err := s.clusterConfiguration.GetAuthPreference(ctx)
+			if err != nil {
+				return 0, trace.Wrap(err)
+			}
+			authPref = ap
+		}
+
+		uid, maybeStale, err := s.createNewStableUNIXUser(ctx, username, authPref)
+		if err != nil {
+			if errors.Is(err, backend.ErrConditionFailed) {
+				continue
+			}
+			// if the readOnlyCache is a bit stale we might end up not checking
+			// the revision of the cached AuthPreference because the feature
+			// might appear to be disabled or the range might be exhausted;
+			// since the readOnlyCache is fed by the auth cache, it could
+			// potentially be stale for longer than the 1600ms of its internal
+			// FnCache, so we can't help but fall back to a backend fetch in
+			// that case
+			if attempt == 0 && maybeStale {
+				continue
+			}
+			return 0, trace.Wrap(err)
+		}
+
+		// TODO(espadolini): emit an audit log event with the username and UID
```

Comment:
Assuming this is out of scope for this PR

Reply:
Yes, it's addressed in #51102.